### PR TITLE
Add gRPC devui test into http-advanced module

### PR DESCRIPTION
**Pull request by kshpak**
quarkus-qe/quarkus-test-plans#33

Please message me to present the logic (if needed)

### CODE REVIEW

**Review by Sgitario — APPROVED** [review]

File: http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/DevModeGrpcIntegrationIT.java [hunk 1]
```diff
@@ -1,33 +1,89 @@
...
     static final String NAME = "QE";
 
+    static final int UNARY_ID = 1;
+    static final int SERVER_STREAM_ID = 2;
+    static final int CLIENT_STREAM_ID = 3;
+    static final int BIDIRECTIONAL_STREAM_ID = 4;
+    static final int CLIENT_STREAM_NAMES_SUBMITTED_COUNT = 2;
+    static final int SERVER_STREAM_MESSAGES_RETURNED_COUNT = 5;
+
+    private List<String> grpcSocketMessages = Arrays.asList(
```
Comment (pjgg) [inline http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/DevModeGrpcIntegrationIT.java:58]: This variable could be final

File: http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/DevModeGrpcIntegrationIT.java [hunk 2]
```diff
@@ -40,4 +96,91 @@ public void testGrpcAsClient() {
...
+        for (String message : grpcSocketMessages) {
+            webSocket.send(message);
+        }
+        await().atMost(60, TimeUnit.SECONDS).until(() -> isGrpcStreamsCompleted());
+        checkGrpcDevUiOutput(GrpcWebSocketListener.serviceOutputMessagesMap);
+
+    }
+
+    @Test
+    public void testGrpcDevUIServicesView() throws JsonProcessingException {
```
Comment (pjgg) [inline http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/DevModeGrpcIntegrationIT.java:111]: JsonProcessingException is never thrown

File: http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/DevModeGrpcIntegrationIT.java [hunk 3]
```diff
@@ -40,4 +96,91 @@ public void testGrpcAsClient() {
...
+                    Log.info("Testing gRPC Server stream service...");
+                    messages = getMessagesByStreamId(streamId);
+                    int actualMessagesReturnedCount = Collections.frequency(messages, "Hello Five Times");
+                    assertEquals(SERVER_STREAM_MESSAGES_RETURNED_COUNT, actualMessagesReturnedCount,
+                            "Server Stream service FAILED");
+                    break;
+                case (CLIENT_STREAM_ID):
+                    Log.info("Testing gRPC Client stream service...");
+                    messages = getMessagesByStreamId(streamId);
+                    assertEquals(messages.contains("Total names submitted: " + CLIENT_STREAM_NAMES_SUBMITTED_COUNT),
```
Comment (pjgg) [inline http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/DevModeGrpcIntegrationIT.java:144]: We could use assertTrue for `message.contains`

File: http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/DevModeGrpcIntegrationIT.java [hunk 4]
```diff
@@ -40,4 +96,91 @@ public void testGrpcAsClient() {
...
+        return messages;
+    }
+
+    private boolean isGrpcStreamsCompleted() {
+        final int expectedCompletedStreams = 4;
+        long completedStreams = GrpcWebSocketListener.serviceOutputMessagesMap.entrySet()
+                .stream().map(entry -> entry.getValue())
+                .filter(list -> list.get(list.size() - 1).contains("COMPLETED"))
+                .count();
+        if (completedStreams == expectedCompletedStreams) {
```
Comment (pjgg) [inline http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/DevModeGrpcIntegrationIT.java:181]: This "if/else" statement could simplified by some thing like:
```
 return completedStreams == expectedCompletedStreams;
```

File: http/http-advanced/src/main/java/io/quarkus/ts/http/advanced/GrpcWebSocketListener.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+import okhttp3.WebSocket;
+import okhttp3.WebSocketListener;
+
+public class GrpcWebSocketListener extends WebSocketListener {
+
+    static Map<Integer, List<String>> serviceOutputMessagesMap = new ConcurrentHashMap<>();
+    private ObjectMapper objectMapper = new ObjectMapper();
```
Comment (pjgg) [inline http/http-advanced/src/main/java/io/quarkus/ts/http/advanced/GrpcWebSocketListener.java:18]: this object could final

File: http/http-advanced/src/main/java/io/quarkus/ts/http/advanced/GrpcWebSocketListener.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+    @Override
+    public void onMessage(WebSocket webSocket, String text) {
+        try {
+            ObjectNode node = objectMapper.readValue(text, ObjectNode.class);
+            int id = node.get("id").asInt();
+            List<String> list = serviceOutputMessagesMap.computeIfAbsent(id, k -> new ArrayList<>());
+            synchronized (list) {
```
Comment (pjgg) [inline http/http-advanced/src/main/java/io/quarkus/ts/http/advanced/GrpcWebSocketListener.java:26]: Maybe with a threadSafe data structure as `BlockingQueue` you could avoid this `synchronized` statement

File: http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/DevModeGrpcIntegrationIT.java [hunk 7]
```diff
@@ -40,4 +96,91 @@ public void testGrpcAsClient() {
...
     }
+
+    @Test
+    public void testGrpcDevUISocket() throws JsonProcessingException {
+        for (String message : grpcSocketMessages) {
+            webSocket.send(message);
+        }
+        await().atMost(60, TimeUnit.SECONDS).until(() -> isGrpcStreamsCompleted());
+        checkGrpcDevUiOutput(GrpcWebSocketListener.serviceOutputMessagesMap);
+
```
Comment (pjgg) [inline http/http-advanced/src/test/java/io/quarkus/ts/http/advanced/DevModeGrpcIntegrationIT.java:107]: this is an extra white space?

**Review by pjgg — CHANGES_REQUESTED** [review]

**Comment by pjgg** [comment]
There are some cosmetic changes, but overall looks good to me.

File: http/http-advanced/src/main/java/io/quarkus/ts/http/advanced/GrpcWebSocketListener.java (hunk 6 above)
Comment (kshpak) [reply http/http-advanced/src/main/java/io/quarkus/ts/http/advanced/GrpcWebSocketListener.java:26]: I left it with synchronized, I think it's more readable with it overall

**Comment by kshpak** [comment]
@pjgg thanks for the review, can you please take a final look?

**Review by pjgg — APPROVED** [review]

**Pull request merged**